constraint/pkg/apis/constraints/apis.go

```
@@ -45,3 +56,123 @@ func GetEnforcementAction(constraint *unstructured.Unstructured) (string, error)
+			if err != nil {
+				continue
+			}
+
+			if pt, ok := enforcementPointMap["name"].(string); ok {
+				if _, ok := actionsForEPs[pt]; !ok && pt != "*" {
+					continue
+				}
+
+				action, found, err := getNestedFieldAsString(scopedEnforcementAction, "action")
+				if err != nil || !found {
+					continue
+				}
+
+				switch pt {
+				case "*":
+					for _, ep := range eps {
+						actionsForEPs[ep][action] = true
+					}
+				default:
+					actionsForEPs[pt][action] = true
+				}
+			}
+		}
+	}
+
+	return actionsForEPs, nil
+}
+
+// Helper function to access nested fields as an array.
+func getNestedFieldAsArray(obj map[string]interface{}, fields ...string) ([]interface{}, bool, error) {
+	value, found, err := unstructured.NestedFieldNoCopy(obj, fields...)
+	if err != nil {
+		return nil, false, err
+	}
+	if !found {
+		return nil, false, nil
+	}
+	if arr, ok := value.([]interface{}); ok {
+		return arr, true, nil
+	}
+	return nil, false, nil
+}
+
+// Helper function to access nested fields as a string.
+func getNestedFieldAsString(obj map[string]interface{}, fields ...string) (string, bool, error) {
+	for _, field := range fields {
+		value, found, err := unstructured.NestedString(obj, field)
+		if err != nil {
+			return "", false, err
+		}
+		if found {
+			return value, true, nil
+		}
+	}
+	return "", false, nil
+}
+
+// Helper function to convert a value to a map[string]interface{}.
+func convertToMapInterface(value interface{}) (map[string]interface{}, error) {
+	if m, ok := value.(map[string]interface{}); ok {
+		return m, nil
+	}
+	return nil, fmt.Errorf("value must be a map[string]interface{}")
+}
+
+// Helper function to convert a value to a []map[string]interface{}.
+func convertToMapSlice(value interface{}) ([]map[string]interface{}, error) {
+	if arr, ok := value.([]interface{}); ok {
+		result := make([]map[string]interface{}, 0, len(arr))
+		for _, v := range arr {
+			if m, ok := v.(map[string]interface{}); ok {
```

This currently skips elements that have values that are not of type map[string]interface{}. Is that intentional? do we need to return an error or log error?

---

Updated the code as per max's comment here - https://github.com/open-policy-agent/frameworks/pull/403/#discussion_r1558699319

---

We probably want to give an error if there is unrecognized data

---

Yeah, we are now returning errors on unrecognized data.

---

I don't see any error returned when v is not of type map[string]interface{}
> if m, ok := v.(map[string]interface{}); ok {

nothing is checking the !ok

---

refactored the function to return proper error.